feat(server): support multiple auth tokens

tessus opened the pull request:
## Description

This change supports the use of multiple authentication tokens.

Example:

```toml
[server]
auth_tokens = [
  "super_secret_token1",
  "super_secret_token2",
]
```

The previously used `AUTH_TOKEN` environment variable can still be used and will be evaluated as well.

## Motivation and Context

Allow to give people access to an otherwise private rustypaste instance.
see https://github.com/orhun/rustypaste/discussions/79

## How Has This Been Tested?

- test cases (cargo test)
- fixtures 
- manual testing on a Linux test VM

## Changelog Entry

````
### Added

- Add support for multiple auth tokens

Example:

```toml
[server]
auth_tokens = [
  "super_secret_token1",
  "super_secret_token2",
]
```

The previously used `AUTH_TOKEN` environment variable can still be used and will be evaluated as well.
````


## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (no code change)
- [ ] Refactor (refactoring production code)
- [ ] Other <!--- (provide information) -->

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly.
- [x] I have formatted the code with [rustfmt](https://github.com/rust-lang/rustfmt).
- [x] I checked the lints with [clippy](https://github.com/rust-lang/rust-clippy).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

orhun commented on fixtures/test-server-auth-multiple-tokens/test.sh:
```diff
@@ -0,0 +1,23 @@
+  result=$(curl -s -F "file=@file" localhost:8000)
+  test "unauthorized" = "$result"
+
+  result=$(curl -s -F "file=@file" -H "Authorization: $auth_token" localhost:8000)
```
Let's also test access with alternative tokens ("token1", "token2" etc.) here.

tessus replied:
sure.

tessus replied:
done

orhun commented on src/config.rs:
```diff
@@ -45,6 +45,8 @@ pub struct ServerConfig {
     pub timeout: Option<Duration>,
     /// Authentication token.
     pub auth_token: Option<String>,
+    /// Authentication tokens.
+    pub auth_tokens: Option<Vec<String>>,
```
How about we deprecate `auth_token` field and only support `auth_tokens`? I think it would help greatly refactoring the code and also having an array is definitely more convenient. What do you think?

tessus replied:
Hah. Yep, I was thinking about this as well. But then I have noticed something interesting: it seems that this config option was never made public. It's not in the example `config.toml`, nor in the README. IMO it is perfectly fine to deprecate it (and get rid of it when you create a `1.0.0` release in the future), but I still think that the env var should be kept. Especially for the Docker folks who like to configure as much as possible with env vars.

You don't even have to mention the deprecation, because this option was never mentioned in te first place either. Well, this only works, if you keep the env var.

In many cases a single auth token _might_ still be the most prevalent use-case, thus keeping the env var could be a good thing. But it's up to you.

orhun replied:
I think it was once mentioned in the changelog and release page which is enough for visibility. Also, this was a requested feature in #14 

We should keep the environment variable for sure. I meant `server.auth_token` -> `server.auth_tokens` and showing a log message if `server.auth_token` is used in the configuration file.

tessus replied:
Yep, this makes perfect sense. I'll add that.

tessus replied:
Done.

tessus commented:
Hmm, sorry for the formatting error. I am using a pre-commit hook, so I don't know why it didn't trigger. Very strange.... and annoying.

I have also added an improvement. The current state only printed deprecation warnings when an action happened. e.g. when the `/` was requested and the deprecated option was read from the config.

I have kept the same logic for the new deprecation warning, but I also added these warnings at server startup.

tessus commented:
Argh! (\*pulling my hair out\*)

I ran `cargo clippy`, which apparently does not include the tests.

orhun commented on fixtures/test-server-auth-multiple-tokens/test.sh:
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env bash
+
+auth_tokens="rustypasteisawesome token1 token2 token4"
```
I think we should use a bash array here:

```bash
auth_tokens=("rustypasteisawesome" "token1" "token2" "token4")
```

tessus replied:
Sure, but why? It doesn't give me any advantage. I am just curious.

I love arrays. The reason why I am careful with them is that macOS comes with an ancient version of bash (3.something). I am using a MacPorts bash, but with newer macOS it is even harder to replace `/bin/bash` even with SIP off.
Thus I usually use `#!/usr/bin/env bash`, but sometimes I forget and use `#!/bin/bash` - and then all hell breaks lose. ;-)
Unless I need associative arrays or want to do fancy array stuff, I just use the above used construct.

But I am more than happy to change it. One array coming up.

tessus replied:
done

orhun commented on src/auth.rs:
```diff
@@ -4,22 +4,41 @@ use actix_web::{error, Error};
+pub fn check(
+    host: &str,
+    headers: &HeaderMap,
+    token: Option<String>,
+    tokens: Option<Vec<String>>,
```
Can we pass a single argument here (only `tokens`) instead?

tessus replied:
This would be a lot easier after the token was removed from the code (release 1.0.0). Otherwise I have to merge token and tokens before passing them into the check function. I can look into it though. In theory this should be easy, but my lack of Rust knowledge makes this is a challenge:

I don't know how to pass a reference of config into a function.

e.g.

```rust
let config = Config::parse(&config_path).expect("failed to parse config");

let tokens = merge_tokens(&config);
```

What is the type of config? I don't know how to declare the function. 

```
fn merge_tokens(config: &what-to-put-here) -> Option<Vec<String>> 
```

But I will look into it. This gives me the opportunity to learn more about Rust.

tessus replied:
I think I figured it out. I'll work on it.

tessus replied:
Done. It was remarkably easy. I think I am starting to understand Rust a bit better.

orhun commented on src/main.rs:
```diff
@@ -48,6 +49,17 @@ fn setup(config_folder: &Path) -> IoResult<(Data<RwLock<Config>>, ServerConfig,
     };
     let config = Config::parse(&config_path).expect("failed to parse config");
     log::trace!("{:#?}", config);
+    if config.server.auth_token.is_some() {
+        log::warn!("[server].auth_token is deprecated, please use [server].auth_tokens");
+    }
+    if config.server.landing_page.is_some() {
+        log::warn!("[server].landing_page is deprecated, please use [landing_page].text");
+    }
+    if config.server.landing_page_content_type.is_some() {
+        log::warn!(
+            "[server].landing_page_content_type is deprecated, please use [landing_page].content_type"
+        );
+    }
```
Maybe extract these into a new function called `warn_deprecation` or something like that. What do you think?

tessus replied:
Sure, but since this ciode will be deleted as soon as the deprecated stuff is removed, I didn't bother. Also I run into the same issue here as described here https://github.com/orhun/rustypaste/pull/84#discussion_r1264681653

But I'll look into it. If you have a quick answer as to what the type of config is, so that I can pass it into a function, I can do this in a few minutes. ;-)

Otherwise it will take a bit to figure this out. I also have a few other things planned for today so I might have to do that tomorrow.

tessus replied:
Hmm, I think it might be `Config`. At least the rust-analyzer says so. I'll run a few tests.

tessus replied:
Hmm, that was easier than expected. Done.

orhun commented on fixtures/test-server-auth-multiple-tokens/test.sh:
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env bash
+
+auth_tokens=("rustypasteisawesome token1 token2 token4")
```
```suggestion
auth_tokens=("rustypasteisawesome" "token1" "token2" "token4")
```

tessus replied:
Oops, my bad. Thanks forcatching this!

orhun commented on src/auth.rs:
```diff
@@ -4,22 +4,31 @@ use actix_web::{error, Error};
-                    auth_header.unwrap_or("none"),
-                );
-                return Err(error::ErrorUnauthorized("unauthorized"));
+pub fn check(host: &str, headers: &HeaderMap, tokens: Option<Vec<String>>) -> Result<(), Error> {
+    if tokens.is_some() {
+        let mut token_found = false;
+        let auth_header = headers
+            .get(AUTHORIZATION)
+            .map(|v| v.to_str().unwrap_or_default())
+            .map(|v| v.split_whitespace().last().unwrap_or_default());
+        if let Some(tokens) = tokens {
+            if !token_found {
+                for token in &tokens {
+                    if auth_header.unwrap_or_default() == token {
+                        token_found = true;
+                        break;
+                    }
+                }
             }
         }
+        if !token_found {
+            log::warn!(
+                "authorization failure for {} (header: {})",
+                host,
+                auth_header.unwrap_or("none"),
+            );
+            return Err(error::ErrorUnauthorized("unauthorized"));
+        }
     }
     Ok(())
 }
```
I think this method can be simplified as:

```rs
pub fn check(host: &str, headers: &HeaderMap, tokens: Option<Vec<String>>) -> Result<(), Error> {
    if let Some(tokens) = tokens {
        let auth_header = headers
            .get(AUTHORIZATION)
            .map(|v| v.to_str().unwrap_or_default())
            .map(|v| v.split_whitespace().last().unwrap_or_default());
        if !tokens.iter().any(|v| v == auth_header.unwrap_or_default()) {
            log::warn!(
                "authorization failure for {} (header: {})",
                host,
                auth_header.unwrap_or("none"),
            );
            return Err(error::ErrorUnauthorized("unauthorized"));
        }
    }
    Ok(())
}
```

tessus replied:
Nice, let me run some tests. e.g. does the `tokens.iter().any` loop stop, when one is found. e.g. I don't want to loop over a huge array, if a token was matched in the beginning of the array already...
This is why I kept my longer construct. I thought about perf.

tessus replied:
Yep, the loop is discontinued as soon as a token is found. Thus this code change is awesome. Makes the code a lot shorter. Nice. Thanks for the tip.

tessus replied:
Done

orhun commented on src/server.rs:
```diff
@@ -306,6 +297,27 @@ pub fn configure_routes(cfg: &mut web::ServiceConfig) {
 }
 
+/// Retrieves all configured tokens.
+#[allow(deprecated)]
+fn get_tokens(config: &Config) -> Option<Vec<String>> {
+    if config.server.auth_token.is_some() || config.server.auth_tokens.is_some() {
+        let mut merged: Vec<String> = vec![];
+        if let Some(tokens) = &config.server.auth_tokens {
+            merged = tokens.clone();
+        }
+        if let Some(token) = &config.server.auth_token {
+            log::warn!("[server].auth_token is deprecated, please use [server].auth_tokens");
+            merged.insert(0, token.to_string());
+        }
+        if let Ok(env_token) = env::var(AUTH_TOKEN_ENV) {
+            merged.insert(0, env_token);
+        }
+        Some(merged)
+    } else {
+        None
+    }
+}
```
Please move this method to make it a part of `Config`. I think calling `config.get_tokens()` is cleaner.

tessus replied:
I'll look into it.

tessus replied:
Done. Btw, I have done the same for the deprecation warning function.

orhun commented on src/config.rs:
```diff
@@ -107,6 +112,42 @@ impl Config {
     }
+    /// Retrieves all configured tokens.
+    #[allow(deprecated)]
+    pub fn get_tokens(&self) -> Option<Vec<String>> {
+        if self.server.auth_token.is_some() || self.server.auth_tokens.is_some() {
+            let mut merged: Vec<String> = vec![];
+            if let Some(tokens) = &self.server.auth_tokens {
+                merged = tokens.clone();
+            }
+            if let Some(token) = &self.server.auth_token {
+                log::warn!("[server].auth_token is deprecated, please use [server].auth_tokens");
+                merged.insert(0, token.to_string());
+            }
+            if let Ok(env_token) = env::var(AUTH_TOKEN_ENV) {
+                merged.insert(0, env_token);
+            }
+            if !merged.is_empty() {
+                return Some(merged);
+            }
+        }
+        None
+    }
```
This function can be refactored as:

```rs
    pub fn get_tokens(&self) -> Option<Vec<String>> {
        let mut tokens = self.server.auth_tokens.clone().unwrap_or_default();
        if let Some(token) = &self.server.auth_token {
            log::warn!("[server].auth_token is deprecated, please use [server].auth_tokens");
            tokens.insert(0, token.to_string());
        }
        if let Ok(env_token) = env::var(AUTH_TOKEN_ENV) {
            tokens.insert(0, env_token);
        }
        (!tokens.is_empty()).then_some(tokens)
    }
```

tessus replied:
Wow, I think I still have to get more into the spirit of Rust.

orhun reviewed (approved):
Thank you for your contribution!

tessus commented:
Thank you for your help and guidance. I always learn something new thanks to you.

orhun commented:
Anytime!
